src/nvim/ex_docmd.c

```diff
@@ -9625,20 +9625,40 @@ static void ex_folddo(exarg_T *eap)
 static void ex_terminal(exarg_T *eap)
 {
   char *name = (char *)p_sh;  // Default to 'shell' if {cmd} is not given.
+  char ex_cmd[1024] = { 0 };
   bool mustfree = false;
-  char *lquote = "['";
-  char *rquote = "']";
 
+  // We will call termopen() with shell as an arglist if not given a {cmd}.
   if (*eap->arg != NUL) {
     name = (char *)vim_strsave_escaped(eap->arg, (char_u *)"\"\\");
     mustfree = true;
-    lquote = rquote = "\"";
+    snprintf(ex_cmd, sizeof(ex_cmd),
+             ":enew%s | call termopen(\"%s\") | startinsert",
+             eap->forceit ? "!" : "", name);
+  } else {
+    char **argv = build_argv((char *)p_sh);
+    char **p = argv;
+    char tempstring[512] = { 0 };
+    char shell_argv[512] = { 0 };
+
+    if (*p != NULL) {
+      snprintf(tempstring, sizeof(tempstring), "\"%s\"", *p);
+      xstrlcat(shell_argv, tempstring, sizeof(shell_argv));
+      p++;
+    }
+
+    while (*p != NULL) {
+      snprintf(tempstring, sizeof(tempstring), ", \"%s\"", *p);
+      xstrlcat(shell_argv, tempstring, sizeof(shell_argv));
+      p++;
+    }
```

Comment:
We could remove the previous `if` block, if we use `shell_argv+2` in the following `snprintf()`.